src/main/java/com/palantir/common/streams/ConcurrentStreams.java

```diff
@@ -0,0 +1,120 @@
+package com.palantir.common.streams;
+
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.immutables.value.Value;
+
+public class ConcurrentStreams {
+
+    private ConcurrentStreams() {}
+
+    private static final TaskCanceledException TASK_CANCELED_EXCEPTION = new TaskCanceledException();
+
+    /**
+     * Runs a map function over all elements in a list with a provided executor and concurrency level.
+     *
+     * @param values The elements to be mapped.
+     * @param mapper The function that maps the elements.
+     * @param executor The executor that runs the concurrent operations.
+     * @param concurrency The max number of operations to be run in parallel. Note that this will
+     *        not check the size of the underlying executor, so ideally the executor should have at least
+     *        as many threads as this value.
+     * @return a stream of mapped elements from the provided list.
+     */
+    public static <T, S> Stream<S> map(
+            List<T> values, Function<T, S> mapper, Executor executor, int concurrency) {
```

Comment:
Why not let the executor handle the concurrency?

Reply:
So that we can used a shared executor and limit a user to some subset of threads. For example: https://github.com/palantir/atlasdb/blob/develop/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java#L701.

Reply:
for example, in the project I've been working on internally, we want to limit the amount of stuff a given requestor keeps in memory at one time.